crates/core/src/buc/store/file.rs

```diff
@@ -0,0 +1,492 @@
+		let Ok(url) = Url::parse(url) else {
+			return Ok(None);
+		};
+
+		if url.scheme() != "file" {
+			return Ok(None);
+		}
+
+		// Get the path from the URL
+		let path = url.path();
+
+		// Create a PathBuf from the path
+		let path_buf = PathBuf::from(path);
+
+		// Check if the path is allowed
+		if !is_path_allowed(&path_buf) {
+			return Err(Error::FileAccessDenied(path.to_string()));
+		}
+
+		// Check if the path exists
+		if !path_buf.exists() {
+			// Create directory and its parents if they don't exist
+			std::fs::create_dir_all(&path_buf).map_err(|e| {
+				Error::InvalidBucketUrl(format!("Failed to create directory {}: {}", path, e))
+			})?;
+		} else if !path_buf.is_dir() {
+			// If path exists but is not a directory, return an error
+			return Err(Error::InvalidBucketUrl(format!("Path is not a directory: {}", path)));
+		}
+
+		Ok(Some(FileStoreOptions {
+			root: Key::from(path.to_string()),
+		}))
+	}
+
+	/// Check if a path exists on disk
+	async fn path_exists(path: &OsPath) -> Result<bool, String> {
+		tokio::fs::try_exists(path)
+			.await
+			.map_err(|e| format!("Failed to check if path exists: {}", e))
+	}
+
+	/// Convert a Path to an OsPath, checking against the allowlist
+	fn to_os_path(&self, path: &Key) -> Result<PathBuf, String> {
+		let root = PathBuf::from(self.options.root.as_str());
+
+		// First canonicalize the root (which should exist)
+		let canonical_root = std::fs::canonicalize(&root)
+			.map_err(|e| format!("Failed to canonicalize root path: {}", e))?;
+
+		// Get the relative path components
+		let relative_path = path.as_str().trim_start_matches('/');
+
+		// Combine the canonical root with the relative path
+		let full_path = canonical_root.join(relative_path);
+
+		// Verify the path is within the allowlist without canonicalizing
+		// Since we're starting from a canonicalized root, the path should be valid
+		if !is_path_allowed(&full_path) {
+			return Err(format!("Path is not in the allowlist: {}", full_path.display()));
+		}
+
+		Ok(full_path)
+	}
+
+	/// Create parent directories for a path if they don't exist
+	async fn ensure_parent_dirs(path: &OsPath) -> Result<(), String> {
+		if let Some(parent) = path.parent() {
+			tokio::fs::create_dir_all(parent)
+				.await
+				.map_err(|e| format!("Failed to create directories: {}", e))?;
+		}
+		Ok(())
+	}
+}
+
+/// Check if a path is allowed according to the allowlist
+fn is_path_allowed(path: &std::path::Path) -> bool {
+	// If the allowlist is empty, nothing is allowed
+	if FILE_ALLOWLIST.is_empty() {
+		return false;
+	}
+
+	// Check if the path is within any of the allowed paths
+	FILE_ALLOWLIST.iter().any(|allowed| {
+		if path.starts_with(allowed) {
+			return true;
+		}
+
+		// Handle case sensitivity on Windows
+		#[cfg(windows)]
+		if path
+			.to_string_lossy()
+			.to_lowercase()
+			.starts_with(&allowed.to_string_lossy().to_lowercase())
+		{
+			return true;
+		}
+
+		false
+	})
+}
```

Comment:
There are a bunch of issues here, first the allow list can be circumvented by having the same prefix and then using `..` to move out of the prefix. Two: not all paths are case insensitive on windows and not all paths are case sensitive on linux. For both of these the way to ensure correct behavior is to always canonicalize the full path and then check against that canonicalized path. 

Three: `Path::exists` is a blocking operation, the tokio way to do it is via `tokio::fs::metadata(path).is_ok()`. Note that canonicalize will also fail if the path does not exist.

Reply:
These should be resolved now. On point two, I opted to make the check case sensitive always now, as the user controls what they put into the file allowlist, and into the bucket's backend URL, that should not be an issue. Happy to hear your thoughts if you have a better solution. Canonicalized are not a solution here, as the path we're checking for may not yet exist

Reply:
Making them both case sensitive makes the behavior different on different file systems which is maybe not what we want. Another solution might be to force a single case on a path and always store files with a lower case path. That should make behavior consistent on all file systems.

Reply:
What about https://crates.io/crates/path-clean or https://crates.io/crates/path-absolutize?

Reply:
File-backend paths are forcefully lowercased by default now, which can be disabled by setting `?lowercase_paths=false`. See [0ea9059](https://github.com/surrealdb/surrealdb/pull/5701/commits/0ea905957ac872659796d344d021502799bc5c60)